### PR TITLE
feat(deploy_ali_cdn): support Alibaba Cloud CDN deployment

### PR DESCRIPTION
**Pull request by PMExtra**
# API Documents
## zh-cn
* https://api.aliyun.com/document/Cdn/2018-05-10/SetCdnDomainSSLCertificate
* https://help.aliyun.com/zh/sdk/product-overview/rpc-mechanism

## en
* https://www.alibabacloud.com/help/en/cdn/developer-reference/api-cdn-2018-05-10-setcdndomainsslcertificate
* https://www.alibabacloud.com/help/en/sdk/product-overview/rpc-mechanism

### CODE REVIEW

File: deploy/ali_cdn.sh [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  if [ -z "$Ali_Key" ] || [ -z "$Ali_Secret" ]; then
+    Ali_Key=""
+    Ali_Secret=""
+    _err "You don't specify aliyun api key and secret yet."
+    return 1
+  fi
+
+  #save the api key and secret to the account conf file.
+  _saveaccountconf_mutable Ali_Key "$Ali_Key"
+  _saveaccountconf_mutable Ali_Secret "$Ali_Secret"
```
Comment (PMExtra) [inline deploy/ali_cdn.sh:40]: Copy from https://github.com/acmesh-official/acme.sh/blob/377a37e4c9c23bb6988fe5f8863f21b19d3e3a40/dnsapi/dns_ali.sh#L13-L24

File: deploy/ali_cdn.sh [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+_url_encode_upper() {
+  encoded=$(_url_encode)
+
+  for match in $(echo "$encoded" | _egrep_o '%..' | sort -u); do
+    upper=$(echo "$match" | _upper_case)
+    encoded=$(echo "$encoded" | sed "s/$match/$upper/g")
+  done
+
+  echo "$encoded"
+}
```
Comment (PMExtra) [inline deploy/ali_cdn.sh:139]: Alibaba Cloud API signature needs parameters encoded in upper-case.

`_url_encode` function from `acme.sh` can only encode with lower-case.

We have an `ali_urlencode` function from `dns_ali.sh` but it does not support multi-line strings.

Maybe we can add an argument to `_url_encode` function to make it support upper-case? (it only needs a very simple change in the `*)` case).

File: deploy/ali_cdn.sh [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+_ali_nonce() {
+  #_head_n 1 </dev/urandom | _digest "sha256" hex | cut -c 1-31
+  #Not so good...
+  date +"%s%N" | sed 's/%N//g'
+}
+
+_timestamp() {
+  date -u +"%Y-%m-%dT%H%%3A%M%%3A%SZ"
+}
```
Comment (PMExtra) [inline deploy/ali_cdn.sh:127]: Copy from https://github.com/acmesh-official/acme.sh/blob/3.0.7/dnsapi/dns_ali.sh

File: deploy/ali_cdn.sh [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  fi
+
+  if [ -z "$ign" ]; then
+    message="$(echo "$response" | _egrep_o "\"Message\":\"[^\"]*\"" | cut -d : -f 2 | tr -d \")"
+    if [ "$message" ]; then
+      _err "$message"
+      return 1
+    fi
+  fi
+}
```
Comment (PMExtra) [inline deploy/ali_cdn.sh:100]: Reference https://github.com/acmesh-official/acme.sh/blob/3.0.7/dnsapi/dns_ali.sh

However, the original method was hard coded with the `GET` method.

So, I made a new argument to specify the method.

Notice that Alibaba Cloud API supports passing the parameters by either query or body. (https://help.aliyun.com/zh/sdk/product-overview/rpc-mechanism#section-9x3-wo3-8l9)

But we must sign with all the parameters, in alphabetical order, regardless of where they come from.

So, I didn't support the post-payload yet.

File: deploy/ali_cdn.sh (hunk 2 above)
Comment (PMExtra) [reply deploy/ali_cdn.sh:139]: https://github.com/acmesh-official/acme.sh/blob/0d8a314bcf32c7705f0be11527d34d3b4ce0fa79/acme.sh#L884-L887

Can we change it to:
```bash
 *) 
   case "$1" in
   '[uU]*') # for upper-case
      printf '%%%s' "$_hex_code" 
      ;;
   *) # default to lower-case
      printf '%%%s' "$_hex_code" | _upper_case
      ;;
   ;;
```

**Comment by iflamed** [comment]
> Solve #1461

How to use this deploy script?

**Comment by PMExtra** [comment]
@iflamed 

```
# This deployment required following variables
# export Ali_Key="ALIACCESSKEY"
# export Ali_Secret="ALISECRETKEY"
# export DEPLOY_ALI_CDN_DOMAIN="cdn.example.com"
# If you have more than one domain, just
# export DEPLOY_ALI_CDN_DOMAIN="cdn1.example.com cdn2.example.com"
#
# The credentials are shared with all domains, also shared with dns_ali api
```

For example:

`docker exec -e Ali_Key=xxxx -e Ali_Secret=xxxx -e DEPLOY_ALI_CDN_DOMAIN=cdn.example.com acme.sh --deploy --deploy-hook ali_cdn -d example.com`

**Comment by PMExtra** [comment]
@iflamed wiki: https://github.com/acmesh-official/acme.sh/wiki/deployhooks#33-deploy-your-certificate-to-alibaba-cloud-cdn-aliyun

**Comment by iflamed** [comment]
> @iflamed wiki: https://github.com/acmesh-official/acme.sh/wiki/deployhooks#33-deploy-your-certificate-to-alibaba-cloud-cdn-aliyun

Hope this PR merged ASAP.

**Comment by jtwangfos** [comment]
Great job man, thanks a log. 3>

**Comment by ShirasawaSama** [comment]
@PMExtra Can DCDN be supported? Thanks.

是否可以支持DCDN？

**Comment by PMExtra** [comment]
@ShirasawaSama 目前没有支持，但只要把这一行代码的 `Cdn` 改成 `Dcdn` 应该就行了，你暂且可以复制一份改一下先用着，回头我考虑下加个设置来控制。

https://github.com/acmesh-official/acme.sh/blob/945b7de76fe1e4f136418e05cd80b2b3fc984ec7/deploy/ali_cdn.sh#L145

**Comment by PMExtra** [comment]
@ShirasawaSama 哦，还漏了API版本，一共改两处，一个是 `Cdn` 改 `Dcdn`，另一个是 `2018-05-10` 改 `2018-01-15`

https://github.com/acmesh-official/acme.sh/blob/945b7de76fe1e4f136418e05cd80b2b3fc984ec7/deploy/ali_cdn.sh#L142-L157

**Comment by PMExtra** [comment]
@anjia0532 我认为这样是增加了复杂性，各个产品自己的接口还是需要调用，只是在调用接口的时候，传递一个证书payload，还是传递一个cas id的区别而已。既然都用acme接管证书管理了，就没必要再去用cas。

File: deploy/ali_cdn.sh [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  date +"%s%N" | sed 's/%N//g'
+}
+
+_timestamp() {
+  date -u +"%Y-%m-%dT%H%%3A%M%%3A%SZ"
+}
+
+# stdin stdout
+_url_encode_upper() {
+  encoded=$(_url_encode)
```
Comment (Neilpang) [inline deploy/ali_cdn.sh:131]: 能不能把 encoded 直接全部 _upper_case 一下 ?

Comment (PMExtra) [reply deploy/ali_cdn.sh:131]: 不可以，原始输入是 base64 ，要区分大小写的。
阿里云要求是特殊字符转义的 `%XX` 必须是大写的 HEX ，否则签名验证通不过。
但不能把 base64 给全转成大写。

其实就是为了处理 `+`, `/`, `=` 和 `\n` 字符对应的转义，其余字符必须保持原样。

**Comment by anjia0532** [comment]
> @anjia0532 我认为这样是增加了复杂性，各个产品自己的接口还是需要调用，只是在调用接口的时候，传递一个证书payload，还是传递一个cas id的区别而已。既然都用acme接管证书管理了，就没必要再去用cas。

确实cas增加了复杂性，但是好处是 1. 可以付费使用部署功能（支持部署到腾讯云，aws，华为云，以及阿里云自身云产品），2. 操作主动吊销泄露的证书更方便。3. 方便删除过期证书。 4. 支持从cas下载证书（具体产品里的证书不支持下载）

当然分开写确实更小而美一些。

**Comment by PMExtra** [comment]
1. 既然你都手动操作部署了，那你不如手动上传证书。acme为的是无人值守，这些手动部署的需求我认为不在考虑范围内；
2. acme上传到cas的证书应该是不能通过cas来吊销的，如果需要吊销仍然是必须通过acme操作；
3. 不部署到cas，就不需要删除过期证书了；
4. 从cas下载证书这也是一个人工操作，再次强调使用acme为的是无人值守，既然你都手动操作了，不如手动上传。

**Comment by iflamed** [comment]
> > @iflamed
> > ```
> > # If you have more than one domain, just
> > # export DEPLOY_ALI_CDN_DOMAIN="cdn1.example.com cdn2.example.com"
> > #
> > # The credentials are shared with all domains, also shared with dns_ali api
> > ```
> 
> Dose it support muti-domains? I found it is said that:
> 
> > The accelerated domain name for which you want to configure the SSL certificate. The type of request supported by the domain name must be HTTPS. You can specify only one domain name in each request.
> 
> https://www.alibabacloud.com/help/en/cdn/developer-reference/api-cdn-2018-05-10-setcdndomainsslcertificate#:~:text=You%20can%20specify%20only%20one%20domain%20name%20in%20each%20request.
> 
> And, there is an API that supports muti-domains:
> 
> https://www.alibabacloud.com/help/en/cdn/developer-reference/api-cdn-2018-05-10-batchsetcdndomainservercertificate

I haven't tested it yet.

**Comment by PMExtra** [comment]
@yikuo123 @iflamed

It supports multiple domains by a loop. It will call the `SetCdnDomainSSLCertificate` API for each domain.

**Comment by ShirasawaSama** [comment]
@PMExtra 现在如果需要使用dcdn的话还是需要手动复制一份脚本吗？是否考虑过添加环境变量的形式区分是否为dcdn或者是正常cdn？又或者说再加一个ali_dcdn.sh脚本？

Now if I need to use dcdn, do I still need to copy the script manually? Have you considered adding an environment variable to determine if it's a dcdn deployment or a normal cdn? Or maybe adding another ali_dcdn.sh script?

**Comment by PMExtra** [comment]
@ShirasawaSama 

考虑到可能有人同时需要 CDN 和 DCDN ，所以直接通过环境变量进行区分的方式可能不太合适。

而直接复制一份 `ali_dcdn.sh`，又会导致代码和 `ali_cdn.sh` 有90%以上的重复率。

我个人希望抽出一个类似 `libs/ali_api.sh` 之类的文件，供 `deploy/ali_cdn.sh`, `deploy/ali_dcdn.sh` 复用。甚至 `dnsapi/dns_ali.sh` 也可以重构一下，复用这个 lib 。

除了阿里，包括 awscn / 腾讯云 / 华为云 等等，也都有不止一个服务有证书需求，而其中有大量可以复用的函数。

请问 @Neilpang 对此有什么想法？

**Comment by Neilpang** [comment]
@PMExtra   想法不错.    如果只是有些方法可以直接复用.  可以考虑直接 把对应的文件 source 进来.    可以使用 `_findHook()` 方法来查找对应的文件. 然后, 自己 source 进来就好了.

**Comment by Neilpang** [comment]
@PMExtra 

不要直接写死相对路径.   用`_findHook ` 来查找, 它会返回给你全路径. 

加个libs变动有点大,  内部倒是问题不大,  但是很多第三方环境都集成了acme.sh, 他们可能已经假定了顶层的目录样子.  单加一个目录他们都需要更新安装脚本.  工作量有点大.  如果不更新, 以前工作的hook 可能就不工作了. 比如, 你把很多函数都抽走了.

而且, 所谓的需要公用的云服务商,  两只手都能数完.  对比起来感觉不划算.

咱们最重要的目标之一就是要尽最大努力保持兼容性, 除非外部因素变动.  不能让客户之前能工作的场景升级后爆了. 因为acme的价值和目标就是无人值守.  没有人愿意每天盯着.


简单你就source吧. 你说呢.

**Comment by PMExtra** [comment]
@Neilpang 

没太理解，你的意思是有人会通过第三方脚本来安装和更新 `acme.sh` 吗？

对于使用常规手段进行安装和更新的用户(`acme.sh --install` 和 `acme.sh --upgrade`)，或者直接 `git clone` 的用户，是不会受影响的。

对于不更新脚本的用户，我们的新feature也是不会对它有影响的。

唯一可能受影响的，就是使用非官方建议的方式更新脚本，只拉取特定文件的用户。

那么这种情况下，无论目录结构怎么样，只要新增了文件，它都有可能缺失。我认为这种**超出官方支持范围**，「自作聪明」做精简的行为，就理应承担相应的风险。

或者退一步，我们不对现有的 `dnsapi/dns_ali.sh` 做重构。只对刚刚新增的 `deploy/ali_dns.sh` 和将来新增的功能引入 `libs/ali_api.sh`，这样是否可以接受呢？

**Comment by Neilpang** [comment]
@PMExtra 

终极用户当然大多数都是用脚本安装的.

但是第三方环境, 例如宝塔之类的,  他们是写死路径的,  可能还要对路径设置相关权限.  他们不会用安装脚本.

既然要抽函数避免copy, 那就只保留一份.   保留已有的文件不动 dnsapi/dns_ali.sh.   其他的文件需要引用的, 就source 它.

让新增文件对旧文件保持严格单向依赖.  这样会减少很多麻烦.

**Comment by ShirasawaSama** [comment]
除了dcdn是否还存在其他需要配置证书的场景？如oss、云建站、云函数等？

**Comment by PMExtra** [comment]
> @PMExtra 除了dcdn是否还存在其他需要配置证书的场景？如oss、云建站、云函数等？

是有这些场景的，我在上面已经描述过了。

**Comment by ShirasawaSama** [comment]
> > @PMExtra 除了dcdn是否还存在其他需要配置证书的场景？如oss、云建站、云函数等？
> 
> 是有这些场景的，我在上面已经描述过了。

我个人觉得是否提供一个 `DEPLOY_ALI_SERVICES=cdn,dcdn,oss` 是否能避免出现一大堆 ali_cdn.sh, ali_dcdn.sh, ali_oss.sh 文件的出现？

**Comment by Neilpang** [comment]
>> 我个人觉得是否提供一个 DEPLOY_ALI_SERVICES=cdn,dcdn,oss

这取决于 这几个文件的相似度.  如果仅仅是几个调用参数不一样, 那没问题.

如果几个服务差异比较大,  会导致 all-in-one 脚本太长, 过于复杂而不利于维护.  其中一个除了bug 没人愿意来修,  看一眼都头大.

**Comment by Neilpang** [comment]
>> DEPLOY_ALI_SERVICES=cdn,dcdn,oss

另外, 对于一个证书需要部署到多个服务的场景, 可以使用 多个 `--deploy-hook aaa  --deploy-hook bbb  --deploy-hook ccc` 来解决.

**Pull request merged**